STYLE_GUIDE.md

```diff
@@ -0,0 +1,323 @@
+following function would be rejected and you will be asked to change it to the
+subsequent example.
+
+```haskell
+function :: Foo -> Bar
+function (Foo fooA nestedBar)
+  | null fooA = mempty
+  | otherwise = nestedBar
+```
+
+Use an `if` or a `case` instead, bearing in mind that the indentation
+requirements still apply. 
+
+```haskell
+function :: Foo -> Bar
+function foo =
+  if null (fooA foo) then 
+    mempty
+  else
+    nestedBar foo
+```
+
+Use of the `LambdaCase` language extension is encouraged. Using this extension
+you would rewrite the following function:
+
+```haskell
+foo :: Quux -> String
+foo (Foo xs) =
+  unlines xs
+foo (Bar xs) =
+  unlines xs
+```
+
+To this layout:
+
+```haskell
+foo :: Quux -> String
+foo = \case
+  Foo xs ->
+    unlines xs
+  Bar xs ->
+    unlines xs
+```
+
+## Avoid using primes when naming things.
+
+Using decimal numbers instead is preferred as they are harder to miss when
+reading the code:
+
+Avoid: `x'`,`x''`.
+
+Instead: `x0`, `x1`.
+
+## Use typeclass constraints where possible.
+
+This works better with how Hedgehog tests are normally written and allows for
+more general use, and leaning on parametricity is always helpful.
+
+Avoid:
+```haskell
+... :: ... -> PropertyT IO ()
+--
+... :: ... -> Gen a
+```
+
+Instead:
+```haskell
+... :: MonadTest m => ... -> m ()
+--
+... :: MonadGen g => ... -> g a
+```
+
+
+If you require `IO` within the `m` then use the `MonadIO` constraint:
+
+Instead:
+```haskell
+... :: (MonadIO m, MonadTest m) => ... -> m ()
+```
+
+## Module declarations and imports
+
+The general order of module declarations is:
+
+1. GHC/HADDOCK options pragmas
+2. LANGUAGE pragmas
+3. Module haddock documentation
+4. Module declaration
+
+Any change to the layout of imports should aim for the minimal number of changes
+to lines and spacing. Adding a single function to an import list and then
+applying `stylish-haskell` can result in a huge diff when it comes to reviewing
+the code. So try to use it sparingly.
+
+The two space indentation rule still applies for module headers:
+
+```haskell
+{-# OPTIONS_HADDOCK not-home #-}
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE TypeFamilies #-}
+module Hedgehog.Internal.Source (
+    LineNo(..)
+  , ColumnNo(..)
+  , Span(..)
+  , getCaller
+
+  -- * Re-exports from "GHC.Stack"
+  , CallStack
+  , HasCallStack
+  , callStack
+  , withFrozenCallStack
+  ) where
+```
+
+ALWAYS USE EXPLICIT IMPORT LISTS!
+
+Do not:
+
+```haskell
+import GHC.Conc
+```
+
+Use qualified imports and make the alias something easily identifiable, avoid
+single letter aliases.
+
+```haskell
+import qualified GHC.Conc as Conc
+```
+
+If necessary you can include periods in the alias name:
+
+```haskell
+import qualified Text.Read as Base.Text
+import qualified Data.Text as Data.Text
+```
```

Comment:
I'm unsure of this as an example, as I would probably use:
```haskell
import qualified Text.Read as Read
import qualified Data.Text as Text
```

It is cool to use periods in qualified names though, maybe this is a better example?

```haskell
import qualified Data.ByteString as ByteString
import qualified Data.ByteString.Lazy as ByteString.Lazy
import qualified Data.Text as Text
import qualified Data.Text.Lazy as Text.Lazy
```

In cases where the API doesn't overlap I think this is better:

```haskell
import qualified Control.Monad.Trans.State.Lazy as Lazy
import qualified Control.Monad.Trans.State.Strict as Strict
import qualified Control.Monad.Trans.Writer.Lazy as Lazy
import qualified Control.Monad.Trans.Writer.Strict as Strict
import qualified Data.ByteString as Strict
import qualified Data.ByteString.Lazy as Lazy
```

Reply:
Cool beans, I can include all the things :+1: